src/main/java/com/redhat/devtools/lsp4ij/features/documentSymbol/LSPBreadcrumbsRefreshListener.java

```diff
@@ -0,0 +1,111 @@
+    private static final Key<Boolean> NEEDS_RESTART = Key.create(LSPBreadcrumbsRefreshListener.class + ".NEEDS_RESTART");
+
+    @Override
+    @Nullable
+    public Object execute(@NotNull Project project, @NotNull Continuation<? super Unit> continuation) {
+        LanguageServerLifecycleManager.getInstance(project).addLanguageServerLifecycleListener(this);
+        return null;
+    }
+
+    @Override
+    public void handleStatusChanged(LanguageServerWrapper languageServer) {
+        if (languageServer.getServerStatus() == ServerStatus.started) {
+            Project project = languageServer.getProject();
+            for (VirtualFile virtualFile : FileEditorManager.getInstance(project).getOpenFiles()) {
+                PsiFile file = LSPIJUtils.getPsiFile(virtualFile, project);
+                if ((file != null) &&
+                    languageServer.getClientFeatures().getBreadcrumbsFeature().isSupported(file) &&
+                    languageServer.getClientFeatures().getBreadcrumbsFeature().isEnabled(file)) {
+                    for (Editor editor : LSPIJUtils.editorsForFile(virtualFile, project)) {
+                        BreadcrumbsPanel breadcrumbsComponent = BreadcrumbsPanel.getBreadcrumbsComponent(editor);
+                        if (breadcrumbsComponent != null) {
+                            // Only add a listener to the editor once
+                            if (editor.getUserData(ADDED_BREADCRUMB_LISTENER) == null) {
+                                breadcrumbsComponent.addBreadcrumbListener(new LSPBreadcrumbListener(file, editor), breadcrumbsComponent);
+                                editor.putUserData(ADDED_BREADCRUMB_LISTENER, true);
+                            }
+
+                            // Denote the file as needing restart and queue an update to its breadcrumbs
+                            file.putUserData(NEEDS_RESTART, true);
+                            ApplicationManager.getApplication().invokeLater(breadcrumbsComponent::queueUpdate);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void handleLSPMessage(Message message, MessageConsumer consumer, LanguageServerWrapper languageServer) {
+    }
+
+    @Override
+    public void handleError(LanguageServerWrapper languageServer, Throwable exception) {
+    }
+
+    @Override
+    public void dispose() {
+    }
+
+    private record LSPBreadcrumbListener(PsiFile file, Editor editor) implements BreadcrumbListener {
+        private LSPBreadcrumbListener(@NotNull PsiFile file, @NotNull Editor editor) {
+            this.file = file;
+            this.editor = editor;
+        }
+
+        @Override
+        public void breadcrumbsChanged(@NotNull Iterable<? extends Crumb> crumbs) {
+            // If the file is still valid and needs restart, clear that flag and schedule restart
+            if (file.isValid() && (file.getUserData(NEEDS_RESTART) == Boolean.TRUE)) {
```

Comment:
We know that we have new breadcrumbs, so if the file is still valid and is marked as needing analysis restarted, clear the "needs restart" flag and schedule analysis restart. That will result in `StickyLinesPass` being run and new sticky lines being collected based on the new breadcrumbs.